### PR TITLE
New feature: Auto detect if file has heading

### PR DESCRIPTION
**Pull request by susgo**
implements new function for auto detection of heading property

**Motivation**
We use a data preview of 5 - 10 lines where customers get the possibility to change the auto detected delimiter and can check if csv file has heading or not (this is necessary for further data processing steps) For the preview we use the limitation functionality so that we not have to process whole file. For the preselection on preview step I would like to auto detect if the file has a heading. (Customer can change it, if software detection is wrong)

**Solution**
Currently the function only checks if the most frequently data type of the heading / first row is string.
The function doesn't change the public properties or parses data newly. Should we do that here? 
Also I could implement a second mode (don't know how to name it) were the data type of each cell in the first row would be compared with the data type getting from the rest of the columns. That should be different in minimum one column to get some more evidence that first row contains headings. What do you think?

Please feel free to close this request if the feature is not needed for others. Also may be there is a better way to implement that or it is already covered by the existing code and I missed it.

### CODE REVIEW

File: src/extensions/DatatypeTrait.php [hunk 1]
```diff
@@ -71,4 +73,41 @@ public function getDatatypes() {
...
+        } else {
+            $firstRow = $this->data[0];
+        }
+
+        $firstRow = array_filter($firstRow);
+        if (empty($firstRow)){
+            return false;
+        }
+
+        $firstRowDatatype = array_map('ParseCsv\enums\DatatypeEnum::getValidTypeFromSample', $firstRow);
```
Comment (Fonata) [inline src/extensions/DatatypeTrait.php]: That line should read something like `array_map(ParseCsv\enums\DatatypeEnum::class . '::getValidTypeFromSample', $firstRow)` to make sure the file is autoloaded and to allow the IDE to complain if class name or namespace change in the future.

Comment (susgo) [reply src/extensions/DatatypeTrait.php]: Okay, I have changed it. Thanks.

**Comment by susgo** [comment]
Tests fail only for php 5.4 but function needs Requires PHP >= 5.5. Can I correct it in any way? Sorry, can search for my own if you have no time.

**Comment by jimeh** [comment]
If I've understood this correctly, all it does is return true if all fields in the first row are determined to be a string.

I appreciate the work that's gone into this feature, but it seems like it won't be very reliable, and I'd argue it might confuse users and make the library more complicated than needed. I'm personally more in favor of leaving it out so developers are forced to deal with it themselves, since we can't reliably deal with it for them.

But I'll leave it up to @Fonata to decide. I just wanted to voice my opinion :)

**Comment by susgo** [comment]
@jimeh You're right. That's why I would prefer to check if datatypes differ between first row and the rest of the column (in a special mode of this function). For our special use-case this will happen, so we can easy check. But yeah, may be this issue is very special requirement. 😊

**Comment by susgo** [comment]
Should I resolve the conflicts or would it be done in case of merging?

**Comment by Fonata** [comment]
I'd like to have this as a logical extension to the type detection to the library. Mmh, we could interpret the extensions folder so it's not always loaded into the class. I'll think about that - and feel free to suggest something, too. This project is a team effort. 

In my point of view, this pull request needs more documentation - PHPDoc only is fine with me.

@susgo don't worry about the conflicts, unless it stops your progress. I'll resolve any remaining conflicts.

**Comment by susgo** [comment]
With composer may be we can solve it with ignoring the extension directory during autoload with `"exclude-from-classmap"` and then add `post-install-cmd` and/or `post-update-cmd`, but I have not tested it yet. What do you think?

**Comment by Fonata** [comment]
Oh, by loading I meant something like different variants of the class: just writing, just pre-2017 parsing, and parsing with datatype and header detection. The objective would be to make learning the library easy, e.g. by IDE completion of just the things the library user needs. I guess that I suggested something counter-productive. Now that I think about it I'm pretty sure that adding more examples will be much more effective than this trait name/directory structure architectural stuff we talked about before.

@susgo I don't think this particular Composer magic is helping. Off-topic: Michael from the Drupal user group will be giving a small presentation about Composer in the next Drupal meetup on March 12.

**Comment by susgo** [comment]
Oh, sorry that I miss-understood. I think, I can not fully follow your thoughts or derive a decision from that - sorry may be it is caused by my english only. I don't like diferent variants of the class. I think we should revert the last commit, because the benefit is less in comparison to the effort we have with recoding and the user has with understanding. Also we have good tests for these features so we would realize when we break the code.

@Fonata I learned a lot about composer in the last weeks because I improved some install things for our project. Seems I have questions no one other has and there is much more to learn. So much power we don't use yet. So I'm very interested in this topic and I have also thought about giving a talk about this next time. Despite everything it is Drupal user group and unfortunately I do not have the time to extend the functionality to fit our homepage needs. So much ideas and pleasure but less time and prio. So if you also interested in that topic, you should go there!

**Comment by susgo** [comment]
Now, I see we have no decision here? What should we do with this "little" issue?

**Comment by Fonata** [comment]
@susgo sorry for  my silence. The current code looks good enough to be in the master, so I'll get it merged. We can improve the PHPDoc explanations and heuristics later.

**Pull request merged**